nltk/translate/nist_score.py

```diff
@@ -93,44 +93,59 @@ def corpus_nist(list_of_references, hypotheses, n=5):
-    p_denominators = Counter() # Key = ngram order, and value = no. of ngram in ref.
-    sysoutput_lengths = Counter() # Key = ngram order, and value = no. of ngram in hyp.
-    hyp_lengths, ref_lengths = 0, 0
-
-    # Iterate through each hypothesis and their corresponding references.
-    for references, hypothesis in zip(list_of_references, hypotheses):
-        # For each order of ngram, calculate the numerator and
-        # denominator for the corpus-level modified precision.
-        for i, _ in enumerate(range(1,n+1)):
-            p_i = modified_precision(references, hypothesis, i)
-            p_numerators[i] += p_i.numerator
-            p_denominators[i] += p_i.denominator
-            # Adds the no. of ngrams in the hypothesis.
-            sysoutput_lengths[i] += len(hypothesis) - (i - 1)
-
-        # Calculate the hypothesis length and the closest reference length.
-        # Adds them to the corpus-level hypothesis and reference counts.
-        hyp_len =  len(hypothesis)
-        hyp_lengths += hyp_len
-        ref_lengths += closest_ref_length(references, hyp_len)
-
-    # Calculate corpus-level brevity penalty.
-    bp = nist_length_penalty(ref_lengths, hyp_lengths)
-
-    # Collects the various precision values for the different ngram orders.
-    p_n = [Fraction(p_numerators[i], p_denominators[i], _normalize=False)
-           for i, _ in enumerate(range(1,n+1))]
-
-    # Eqn 2 in Doddington (2002):
-    # Info(w_1 ... w_n) = log_2 [ (# of occurrences of w_1 ... w_n-1) / (# of occurrences of w_1 ... w_n) ]
-    info = [0 if p_n[i].numerator == 0 or p_n[i+1].numerator == 0 # Handles math domain and zero division errors.
-            else math.log(p_n[i].numerator / p_n[i+1].numerator)
-            for i in range(len(p_n)-1)]
-    return sum(info_i/sysoutput_lengths[i] for i, info_i in enumerate(info)) * bp
-
-
-def nist_length_penalty(closest_ref_len, hyp_len):
+    # Collect the ngram coounts from the reference sentences.
+    ngram_freq = Counter()
+    total_reference_words = 0
+    for references in list_of_references: # For each source sent, there's a list of reference sents.
+        for reference in references:
+            # For each order of ngram, count the ngram occurrences.
+            for i in range(1,n+1):
+                ngram_freq.update(ngrams(reference, i))
+            total_reference_words += len(reference)
+
+    # Compute the information weights based on the reference sentences.
+    information_weights = {}
+    for _ngram in ngram_freq: # w_1 ... w_n
+        _mgram = _ngram[:-1] #  w_1 ... w_n-1
+        if _mgram and _mgram in ngram_freq:
+            denominator = ngram_freq[_mgram]
+        else:
+            denominator = total_reference_words
+        # Eqn 2 in Doddington (2002):
+        # Info(w_1 ... w_n) = log_2 [ (# of occurrences of w_1 ... w_n-1) / (# of occurrences of w_1 ... w_n) ]
+        information_weights[_ngram] = -1 * math.log(ngram_freq[_ngram]/denominator) / math.log(2)
```

Comment:
Mathematically, this is the same formula as Eqn2 but I've no idea why did the script https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/mteval-v13a.pl#L546 

Is it because of preventing underflow from log of really low fraction value? 

The line could have been this and yield the same result:

```
information_weights[_ngram] = math.log(denominator/ngram_freq[_ngram], 2)
```

Reply:
Does Perl's log function allow you to specify the base?

Reply:
I think we can stick with the version closer to paper's interpretation without the perl's limitation =)